modules/imgproc/test/test_histograms.cpp

```diff
@@ -1206,14 +1180,82 @@ int CV_CalcHistTest::prepare_test_case( int test_case_idx )
 
 void CV_CalcHistTest::run_func(void)
 {
-    cvCalcHist( images, hist[0], 0, images[CV_MAX_DIM] );
+    int size[CV_MAX_DIM];
+    int hdims = cvGetDims( hist[0]->bins, size);
+    bool huniform = CV_IS_UNIFORM_HIST(hist[0]);
+
+    const float* uranges[CV_MAX_DIM] = {0};
+    const float** hranges = 0;
+
+    if( hist[0]->type & CV_HIST_RANGES_FLAG )
+    {
+        hranges = (const float**)hist[0]->thresh2;
+        if( huniform )
+        {
+            for(int i = 0; i < hdims; i++ )
+                uranges[i] = &hist[0]->thresh[i][0];
+            hranges = uranges;
+        }
+    }
+
+    std::vector<cv::Mat> imagesv(cdims);
+    copy(images.begin(), images.begin() + cdims, imagesv.begin());
+
+    Mat mask = images[CV_MAX_DIM];
+    if( !CV_IS_SPARSE_HIST(hist[0]) )
+    {
+        cv::Mat H = cv::cvarrToMat(hist[0]->bins);
+        if(huniform)
+        {
+            vector<int> emptyChannels;
+            vector<int> hSize(hdims);
+            for(int i = 0; i < hdims; i++)
+                hSize[i] = size[i];
+            vector<float> vranges;
+            if(hranges)
+            {
+                vranges.resize(hdims*2);
+                for(int i = 0; i < hdims; i++ )
+                {
+                    vranges[i*2  ] = hist[0]->thresh[i][0];
+                    vranges[i*2+1] = hist[0]->thresh[i][1];
+                }
+            }
+            cv::calcHist(imagesv, emptyChannels, mask, H, hSize, vranges);
+        }
+        else
+        {
+            cv::calcHist( &imagesv[0], (int)imagesv.size(), 0, mask,
+                          H, cvGetDims(hist[0]->bins), H.size, hranges, huniform );
+        }
+    }
+    else
+    {
+        CvSparseMat* sparsemat = (CvSparseMat*)hist[0]->bins;
+
+        cvZero( hist[0]->bins );
+
+        cv::SparseMat sH;
+        sparsemat->copyToSparseMat(sH);
+
+        cv::calcHist( &imagesv[0], (int)imagesv.size(), 0, mask, sH, sH.dims(),
+                      sH.dims() > 0 ? sH.hdr->size : 0, hranges, huniform, false);
+
+        cv::SparseMatConstIterator it = sH.begin();
+        int nz = (int)sH.nzcount();
+        for(int i = 0; i < nz; i++, ++it )
+        {
+            CV_Assert(it.ptr != NULL);
+            *(float*)cvPtrND(sparsemat, it.node()->idx, 0, -2) = *(const float*)it.ptr;
+        }
+    }
 }
 
 
 static void
-cvTsCalcHist( IplImage** _images, CvHistogram* hist, IplImage* _mask, int* channels )
+cvTsCalcHist( vector<Mat> images, CvHistogram* hist, Mat mask, vector<int> channels )
```

Comment:
`const std::vector<Mat>&`
`const std::vector<int>&`

here and below

Reply:
fixed